src/Features/Core/Portable/Completion/CompletionHelper.cs

```diff
@@ -358,5 +359,109 @@ private static int CompareExpandedItem(CompletionItem item1, PatternMatch match1
+            return false;
+
+            static bool MatchesFilterText(
+                CompletionItem item,
+                string filterText,
+                CompletionTriggerKind initialTriggerKind,
+                CompletionFilterReason filterReason,
+                ImmutableArray<string> recentItems,
+                PatternMatch? patternMatch)
+            {
+                // For the deletion we bake in the core logic for how matching should work.
+                // This way deletion feels the same across all languages that opt into deletion 
+                // as a completion trigger.
+
+                // Specifically, to avoid being too aggressive when matching an item during 
+                // completion, we require that the current filter text be a prefix of the 
+                // item in the list.
+                if (filterReason == CompletionFilterReason.Deletion &&
+                    initialTriggerKind == CompletionTriggerKind.Deletion)
+                {
+                    return item.FilterText.GetCaseInsensitivePrefixLength(filterText) > 0;
+                }
+
+                // If the user hasn't typed anything, and this item was preselected, or was in the
+                // MRU list, then we definitely want to include it.
+                if (filterText.Length == 0)
+                {
+                    if (item.Rules.MatchPriority > MatchPriority.Default)
+                    {
+                        return true;
+                    }
+
+                    if (!recentItems.IsDefault && GetRecentItemIndex(recentItems, item) <= 0)
+                    {
+                        return true;
+                    }
+                }
+
+                // Otherwise, the item matches filter text if a pattern match is returned.
+                return patternMatch != null;
+            }
+
+            static int GetRecentItemIndex(ImmutableArray<string> recentItems, CompletionItem item)
+            {
+                var index = recentItems.IndexOf(item.FilterText);
+                return -index;
+            }
+
+            // If the item didn't match the filter text, we still keep it in the list
+            // if one of two things is true:
+            //  1. The user has typed nothing or only typed a single character.  In this case they might
+            //     have just typed the character to get completion.  Filtering out items
+            //     here is not desirable.
+            //
+            //  2. They brought up completion with ctrl-j or through deletion.  In these
+            //     cases we just always keep all the items in the list.
+            static bool KeepAllItemsInTheList(CompletionTriggerKind initialTriggerKind, string filterText)
+            {
+                return filterText.Length <= 1 ||
+                    initialTriggerKind == CompletionTriggerKind.Invoke ||
+                    initialTriggerKind == CompletionTriggerKind.Deletion;
+            }
```

Comment:
was this logic effectively all unchanged when you moved it down?

Reply:
Yup, all unchanged!